File: packages/compiler/playground/ios/HelloMyStateBag/Diez/Sources/Components/MyStateBag.swift
import Foundation

public final class MyPalette : NSObject, StateBag {
    public var hello: Color

    public init(_ listener: Method?) {
        hello = Color(hue: 0, hslSaturation: 1, lightness: 0.5, alpha: 1)
    }

    public func update(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        // TODO: We should be able to patch instead of replacing here.
        // Is it possible to update UIColors?
        hello = try container.decode(Color.self, forKey: .hello)
    }
}

public final class MyStateBag : NSObject, StateBag {
    public var palette: MyPalette
    public var copy: String
    public var image: Image
    public var fontRegistry: FontRegistry
    public var textStyle: TextStyle
    public var haiku: Haiku
    public var svg: SVG
    public var lottie: Lottie
    var listener: Method? = nil

    private enum CodingKeys: String, CodingKey {
        case palette
        case copy
        case image
        case fontRegistry
        case textStyle
        case haiku
        case svg
        case lottie
    }

    public init(_ listenerIn: Method?) {
        fontRegistry = FontRegistry(withFiles: [
            File(withSrc: "/assets/fonts/Roboto-Black.ttf"),
            File(withSrc: "/assets/fonts/Roboto-BlackItalic.ttf"),
            File(withSrc: "/assets/fonts/Roboto-Bold.ttf"),
            File(withSrc: "/assets/fonts/Roboto-BoldItalic.ttf"),
            File(withSrc: "/assets/fonts/Roboto-Italic.ttf"),
            File(withSrc: "/assets/fonts/Roboto-Light.ttf"),
            File(withSrc: "/assets/fonts/Roboto-LightItalic.ttf"),
            File(withSrc: "/assets/fonts/Roboto-Medium.ttf"),
            File(withSrc: "/assets/fonts/Roboto-MediumItalic.ttf"),
            File(withSrc: "/assets/fonts/Roboto-Regular.ttf"),
            File(withSrc: "/assets/fonts/Roboto-Thin.ttf"),
            File(withSrc: "/assets/fonts/Roboto-ThinItalic.ttf"),
        ])
        palette = MyPalette(listener)
        copy = "Hello Diez"
        image = Image(withFile: File(withSrc: "/assets/images/haiku.jpg"), withWidth: 246, withHeight: 246, withScale: 3)
        textStyle = TextStyle(
            withFont: "Helvetica",
            withFontSize: 50,
            withColor: Color(hue: 0, hslSaturation: 1, lightness: 0.5, alpha: 1)
        )
        listener = listenerIn
        haiku = Haiku(withFile: File(withSrc: "/assets/haiku/animator.html"))
        svg = SVG(withFile: File(withSrc: "/assets/images/rat.svg.html"))
        lottie = Lottie(withFile: File(withSrc: "/assets/lottie/loading-pizza.json"))
    }

    public func update(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        try? container.update(&palette, forKey: .palette)
        try? container.update(&image, forKey: .image)
        try? container.update(&textStyle, forKey: .textStyle)
        try? container.update(&svg, forKey: .svg)
        try? container.update(&haiku, forKey: .haiku)
        try? container.update(&lottie, forKey: .lottie)
        copy = try container.decode(String.self, forKey: .copy)
    }

    public func tap() {
        if (listener == nil) {
            return
        }

        listener!("tap", nil)
    }
}

File: examples/playground/android/diez/src/main/java/ai/haiku/diez/components/MyStateBag.kt
package ai.haiku.diez.components

import android.graphics.Color as CoreColor
import ai.haiku.diez.color.Color
import ai.haiku.diez.puente.Method
import ai.haiku.diez.puente.StateBag
import ai.haiku.diez.color.ColorAdapter
import ai.haiku.diez.color.QualifiedColor
import ai.haiku.diez.file.File
import ai.haiku.diez.haiku.Haiku
import ai.haiku.diez.image.Image
import ai.haiku.diez.image.SVG
import ai.haiku.diez.lottie.Lottie
import ai.haiku.diez.typography.FontRegistry
import ai.haiku.diez.typography.TextStyle

val colorAdapter = ColorAdapter()

data class MyPalette(@QualifiedColor var hello: Color = ColorAdapter.hsla(floatArrayOf(0F, 1F, 0.5F, 1F)))

data class MyStateBag(
    var palette: MyPalette = MyPalette(),
    var copy: String = "Hello Diez",
    var image: Image = Image(File("/assets/images/haiku.jpg"), 246, 246, 3F),
    var fontRegistry: FontRegistry = FontRegistry(arrayOf<File>(
        File("/assets/fonts/Roboto-Black.ttf"),
        File("/assets/fonts/Roboto-BlackItalic.ttf"),
        File("/assets/fonts/Roboto-Bold.ttf"),
        File("/assets/fonts/Roboto-BoldItalic.ttf"),
        File("/assets/fonts/Roboto-Italic.ttf"),
        File("/assets/fonts/Roboto-Light.ttf"),
        File("/assets/fonts/Roboto-LightItalic.ttf"),
        File("/assets/fonts/Roboto-Medium.ttf"),
        File("/assets/fonts/Roboto-MediumItalic.ttf"),
        File("/assets/fonts/Roboto-Regular.ttf"),
        File("/assets/fonts/Roboto-Thin.ttf"),
        File("/assets/fonts/Roboto-ThinItalic.ttf")
    )),
    var textStyle: TextStyle = TextStyle(
        "Helvetica",
        50F,
        ColorAdapter.hsla(floatArrayOf(0F, 1F, 0.5F, 1F))
    ),
    var haiku: Haiku = Haiku("@haiku/taylor-testthang"),
    var svg: SVG = SVG(File("/assets/images/rat.svg")),
    var lottie: Lottie = Lottie(File("/assets/lottie/loading-pizza.json"))
) : StateBag {
    override val adapters = listOf(
        colorAdapter
    )

    override val name = "MyStateBag"

    @Transient
    private var listener: Method? = null

    override fun listen(listener: Method) {
        this.listener = listener
    }

    fun tap() {
        if (this.listener == null) {
            return
        }

        this.listener!!("tap", null)
    }
}
